mitodl/micromasters
# Final grade freezing

#### What are the relevant tickets?
closes #1958 
closes #1959

#### What's this PR do?
Introduces APIs, celery tasks and management commands to freeze final grades for course runs.

#### Where should the reviewer start?
`grades.api`, then `grades.tasks`

#### How should this be manually tested?
to freeze a curse run the entry point is:
in the django shell import `grades.tasks.freeze_course_run_final_grades`

or the management commands `freeze_final_grades`

create cached enrollments, current_grades for all the users you want to considered for grade freeze.
create cached certificate for some of the financial aid course run users: this is needed to test the special case in which the certificate should be considered before the current grade also for financial aid courses.

There is a lot to test in this PR:
1. verify that only the courses with a freeze date in the past can be frozen: for this first try to freeze a course without a freeze date (it should raise `ImproperlyConfigured`) then a course with the freeze date in the future (nothing should happen).
2. verify that all the enrolled users have a final grade (with passed or not according to the current_grade and certificates you created)
3. verify that the course run with an entry as COMPLETE in `FinalGradeRunInfo` are not considered for freeze any more
4. any edge case you have in mind.

Be aware that the task `grades.tasks.freeze_course_run_final_grades` needs to run at least twice to complete the freeze for a course.

#### What GIF best describes this PR or how it makes you feel?
![](https://thearoundtheway.files.wordpress.com/2015/01/frozen.gif)
